src/ompl/geometric/planners/aitstar/AITstar.h

```diff
@@ -0,0 +1,295 @@
+
+            /** \brief Get all outgoing edges of a vertex. */
+            std::vector<aitstar::Edge> getOutgoingEdges(const std::shared_ptr<aitstar::Vertex> &vertex) const;
+
+            /** \brief Check whether all vertices of a set of edges have been processed by the reverse queue. */
+            bool haveAllVerticesBeenProcessed(const std::vector<aitstar::Edge> &edges) const;
+
+            /** \brief Check whether the parent and child vertices of an edge have been processed by the reverse queue.
+             */
+            bool haveAllVerticesBeenProcessed(const aitstar::Edge &edges) const;
+
+            /** \brief Checks whether the cost to come of a goal vertex has been updated and updates the solution if so.
+             */
+            void updateExactSolution();
+
+            /** \brief Checks whether the input vertex is the new best approximate solution and updates the solution in
+             * the problem definition if so. **/
+            void updateApproximateSolution(const std::shared_ptr<aitstar::Vertex> &vertex);
+
+            /** \brief Returns the best cost-to-go-heuristic to any start in the graph. */
+            ompl::base::Cost computeCostToGoToStartHeuristic(const std::shared_ptr<aitstar::Vertex> &vertex) const;
+
+            /** \brief Returns the best cost-to-go-heuristic to any goal in the graph. */
+            ompl::base::Cost computeCostToGoToGoalHeuristic(const std::shared_ptr<aitstar::Vertex> &vertex) const;
+
+            /** \brief Returns the best cost to go to any goal in the graph. */
+            ompl::base::Cost computeCostToGoToGoal(const std::shared_ptr<aitstar::Vertex> &vertex) const;
+
+            /** \brief Returns the best cost to come form the goal of any start. */
+            ompl::base::Cost computeBestCostToComeFromGoalOfAnyStart() const;
+
+            /** \brief Sets the cost to come from to goal of the reverse search to infinity for the whole branch.
+             */
+            void invalidateCostToComeFromGoalOfReverseBranch(const std::shared_ptr<aitstar::Vertex> &vertex);
+
+            /** \brief The increasingly dense sampling-based approximation. */
+            aitstar::ImplicitGraph graph_;
+
+            /** \brief The forward queue. */
+            using EdgeQueue =
+                ompl::BinaryHeap<aitstar::Edge, std::function<bool(const aitstar::Edge &, const aitstar::Edge &)>>;
+            std::unique_ptr<EdgeQueue> forwardQueue_;
+
+            /** \brief The reverse queue. */
+            using KeyVertexPair = std::pair<std::array<ompl::base::Cost, 2u>, std::shared_ptr<aitstar::Vertex>>;
+            using VertexQueue =
+                ompl::BinaryHeap<KeyVertexPair, std::function<bool(const KeyVertexPair &, const KeyVertexPair &)>>;
+            std::unique_ptr<VertexQueue> reverseQueue_;
+
+            /** \brief The id of the current forward search .*/
+            std::shared_ptr<std::size_t> forwardSearchId_;
```

Comment:
does this really need to be a shared pointer? Same question for next two data members.

Reply:
They're needed in the `ImplicitGraph` but I think they can be stack allocated and passed by const reference. Thanks for pointing this out.